### PR TITLE
Improve search of authorized and unauthorized studies

### PR DESCRIPTION
**Pull request by jagnathan**
Fix cBioPortal/cbioportal#9915. Improved search of authorized internal studies. Users can now filter studies based on their read permission (authorized or not authorized) in the search box. This is an optional feature that has to be enabled by settings in portal.properties.

### CODE REVIEW

**Review by pvannierop — COMMENTED** [review]
Hi Jag, thank you for making this! The only comment that I have is that now the selection widget are two checkboxes that allow you to unselect both options which would result in an empty study list. Personally I think it would be better to have a single toggle that turns the UNAUTH studies on or off.

File: src/shared/components/query/filteredSearch/field/CheckboxFilterField.tsx [hunk 1]
```diff
@@ -58,7 +58,7 @@ export const FilterCheckbox: FunctionComponent<FieldProps> = props => {
             <h5>{props.filter.form.label}</h5>
             <div>
                 {options.map((option: FilterFieldOption) => {
-                    const id = `input-${option}`;
+                    const id = `input-${option.displayValue}`;
```
Comment (BasLee) [inline src/shared/components/query/filteredSearch/field/CheckboxFilterField.tsx:61]: Filter can now still be matched to the wrong checkbox, when their `displayValues` are the same. Maybe better to just wrap the checkbox by its label?
```js
                {options.map((option: FilterFieldOption) => {
                    let isChecked = checkedOptions.includes(option.value);
                    return (
                        <div
                            style={{
                                display: 'inline-block',
                                padding: '0 1em 0 0',
                            }}
                        >
                            <label
                                style={{
                                    display: 'inline-block',
                                    padding: '0 0 0 0.2em',
                                }}
                            >
                                <input
                                    type="checkbox"
                                    value={option.displayValue}
                                    checked={isChecked}
                                    onClick={() => {
                                        isChecked = !isChecked;
                                        updatePhrases(option.value, isChecked);
                                        const update = createQueryUpdate(
                                            toRemove,
                                            checkedOptions,
                                            props.filter
                                        );
                                        props.onChange(update);
                                    }}
                                    style={{
                                        display: 'inline-block',
                                    }}
                                />
                                {' '}
                                {option.displayValue}
                            </label>
                        </div>
                    );
                })}
                
```

Comment (jagnathan) [reply src/shared/components/query/filteredSearch/field/CheckboxFilterField.tsx]: now, the id will consist of displayvalue and optionvalue combined to ensure that they are unique. Enwrapped with label.

File: src/shared/components/query/CancerStudySelector.tsx [hunk 2]
```diff
@@ -307,45 +313,104 @@ export default class CancerStudySelector extends React.Component<
...
+                                                                    shownAndAuthorizedStudies.length &&
+                                                                shownAndAuthorizedStudies.length >
+                                                                    0
+                                                            }
+                                                        />
+                                                        <strong>
+                                                            {shownAndSelectedStudies.length ===
+                                                                shownAndAuthorizedStudies.length &&
+                                                            shownAndAuthorizedStudies.length >
+                                                                0
```
Comment (dippindots) [inline src/shared/components/query/CancerStudySelector.tsx:387]: This condition shows twice, maybe we can create a constant for it and find a good name for it. e.g. `authorizedStudiesChecked`?

File: src/shared/components/query/CancerStudySelector.tsx [hunk 3]
```diff
@@ -307,45 +313,104 @@ export default class CancerStudySelector extends React.Component<
...
+                                                                ? `Deselect all authorized studies ${
+                                                                    shownStudies.length <
+                                                                      this.store
+                                                                          .cancerStudies
+                                                                          .result
+                                                                          .length
```
Comment (dippindots) [inline src/shared/components/query/CancerStudySelector.tsx:393]: Just wanted to make sure that we are waiting for this data-fetching process to finish somewhere, right? I can't see it's waiting for it in this component.

File: src/shared/components/query/QueryStore.ts [hunk 4]
```diff
@@ -1450,6 +1450,15 @@ export class QueryStore {
         return new Set(referenceGenomes);
     }
 
+    @computed get readPermissions(): Set<string> {
+        const studies = Array.from(this.treeData.map_node_meta.keys()).filter( s => typeof((s as CancerStudy).readPermission) !== 'undefined' );
+        console.log(studies);
```
Comment (dippindots) [inline src/shared/components/query/QueryStore.ts:1455]: We can delete this log

File: src/shared/components/query/QueryStore.ts [hunk 5]
```diff
@@ -1450,6 +1450,15 @@ export class QueryStore {
...
+        const readPermissions = studies
+            .map(n => (!!((n as CancerStudy).readPermission)).toString())
+            .filter(n => !!n);
```
Comment (dippindots) [inline src/shared/components/query/QueryStore.ts:1458]: Could you add some comments about how to get `readPermissions`? I think it's not very straightforward.

File: src/shared/components/query/StudyListLogic.ts [hunk 6]
```diff
@@ -393,6 +393,42 @@ export class FilteredCancerTreeView {
...
+        });
+        if (
+            shownAndAuthorizedStudies.length === shownAndSelectedStudies.length
+        ) {
+            // deselect
+            updatedSelectableSelectedStudyIds = _.without(
+                this.store.selectableSelectedStudyIds,
+                ...shownAndAuthorizedStudies.map(
+                    (study: CancerStudy) => study.studyId
+                )
```
Comment (dippindots) [inline src/shared/components/query/StudyListLogic.ts:416]: This also shows twice, maybe we can create a constant variable for it
```
const shownAndAuthorizedStudyIds = shownAndAuthorizedStudies.map(
    (study: CancerStudy) => study.studyId
);
```

File: src/shared/lib/query/QueryParser.ts [hunk 7]
```diff
@@ -18,14 +18,18 @@ import {
...
     /**
      * Can be extended with additional search filters
      */
     private readonly _searchFilters: CancerTreeSearchFilter[];
 
-    constructor(referenceGenomes: Set<string>) {
+    constructor(referenceGenomes: Set<string>, readPermissions: Set<string>) {
+        console.log('readPermissions');
+        console.log(readPermissions);
+        console.log(readPermissions.size);
```
Comment (dippindots) [inline src/shared/lib/query/QueryParser.ts:32]: We can remove these logs too.

**Review by dippindots — COMMENTED** [review]
@jagnathan Thanks for the pr, overall it looks good. I put some minor comments there. Feel free to address it when you got a chance.
I think it would be great if we can have some tests for this feature. Do you know if it's easy to add some related tests?

File: src/shared/components/query/CancerStudySelector.tsx (hunk 3 above)
Comment (jagnathan) [reply src/shared/components/query/CancerStudySelector.tsx:393]: default value exists for this.store.cancerStudies.

File: src/shared/lib/query/QueryParser.ts (hunk 7 above)
Comment (jagnathan) [reply src/shared/lib/query/QueryParser.ts:32]: removed.

File: src/shared/components/query/CancerStudySelector.tsx (hunk 2 above)
Comment (jagnathan) [reply src/shared/components/query/CancerStudySelector.tsx:387]: new const called shownStudiesLengthstring has been created

File: src/shared/components/query/QueryStore.ts (hunk 5 above)
Comment (jagnathan) [reply src/shared/components/query/QueryStore.ts]: comment added in commit

**Review by dippindots — APPROVED** [review]
@jagnathan Thanks for addressing all the comments, LGTM. Please try to resolve the conflicts and rebase based on the latest master.

**Comment by jagnathan** [comment]
superseded by new PR https://github.com/cBioPortal/cbioportal-frontend/pull/4603

**Pull request closed without merging**